### ISSUE TITLE
Error when using cgbn_div with more than 512-bit numbers

### ISSUE THREAD

**Issue by peterminea** [NVlabs/CGBN#23]
Hello CGBN team!

I am a software developer interested in Number Theory, and I have been working with CGBN and CUDA on C for roughly two months, making implementations for big numbers, mostly for the 256-bit datatype, using CGBN arithmetic routines like subtraction, bit shifting, multiplication, division, and recently I tried to also use my programs for larger numbers with the BITS parameter set to 512 or 1024/2048.

When using sizes of BITS that are greater than 512 (like 544, 576, 1024, 2048), that are divisible by 32, I get strange errors ("incomplete type not allowed") regarding division arithmetic operations (cgbn_div_rem, cgbn_div, it also happened with cgbn_div_rem_wide). It happens when I want to divide and/or get the remainder for two cgbn_t variables.  It does NOT give this error if BITS is 512 or lower, something is happening when using values of BITS that are divisible by 32 and larger than 512.

An example of an error message, for cgbn_div, is:

/CGBN-master/include/cgbn/core/core.cu(154): error: incomplete type is not allowed
          detected during:
            instantiation of "void cgbn::core_t<env>::dlimbs_approximate(uint32_t *, const uint32_t *) [with env=cgbn::unpadded_t<cgbn_env_t<context_t, 1024U, (cgbn_syncable_t)0>>]" 
/CGBN-master/include/cgbn/core/core_divide_multi.cu(57): here
            instantiation of "void cgbn::core_t<env>::div_wide(uint32_t *, const uint32_t *, const uint32_t *, const uint32_t *, uint32_t) [with env=cgbn::unpadded_t<cgbn_env_t<context_t, 1024U, (cgbn_syncable_t)0>>]" 
/CGBN-master/include/cgbn/core/core_singleton.cu(133): here
            instantiation of "void cgbn::core_singleton_t<env, limbs>::div_wide(uint32_t *, const uint32_t *, const uint32_t *, const uint32_t *, uint32_t) [with env=cgbn::unpadded_t<cgbn_env_t<context_t, 1024U, (cgbn_syncable_t)0>>, limbs=8U]" 
/CGBN-master/include/cgbn/impl_cuda.cu(407): here
            instantiation of "void cgbn_env_t<context_t, bits, syncable>::div(cgbn_env_t<context_t, bits, syncable>::cgbn_t &, const cgbn_env_t<context_t, bits, syncable>::cgbn_t &, const cgbn_env_t<context_t, bits, syncable>::cgbn_t &) const [with context_t=context_t, bits=1024U, syncable=(cgbn_syncable_t)0]" 
/CGBN-master/include/cgbn/cgbn.h(132): here
            instantiation of "void cgbn_div(env_t, env_t::cgbn_t &, const env_t::cgbn_t &, const env_t::cgbn_t &) [with env_t=env_t]" 
modnumspiral.cu(140): here

/CGBN-master/include/cgbn/core/core.cu(158): error: incomplete type is not allowed
          detected during:
            instantiation of "void cgbn::core_t<env>::dlimbs_div_estimate(uint32_t *, const uint32_t *, const uint32_t *) [with env=cgbn::unpadded_t<cgbn_env_t<context_t, 1024U, (cgbn_syncable_t)0>>]" 
/CGBN-master/include/cgbn/core/core_divide_multi.cu(63): here
            instantiation of "void cgbn::core_t<env>::div_wide(uint32_t *, const uint32_t *, const uint32_t *, const uint32_t *, uint32_t) [with env=cgbn::unpadded_t<cgbn_env_t<context_t, 1024U, (cgbn_syncable_t)0>>]" 
/CGBN-master/include/cgbn/core/core_singleton.cu(133): here
            instantiation of "void cgbn::core_singleton_t<env, limbs>::div_wide(uint32_t *, const uint32_t *, const uint32_t *, const uint32_t *, uint32_t) [with env=cgbn::unpadded_t<cgbn_env_t<context_t, 1024U, (cgbn_syncable_t)0>>, limbs=8U]" 
/CGBN-master/include/cgbn/impl_cuda.cu(407): here
            instantiation of "void cgbn_env_t<context_t, bits, syncable>::div(cgbn_env_t<context_t, bits, syncable>::cgbn_t &, const cgbn_env_t<context_t, bits, syncable>::cgbn_t &, const cgbn_env_t<context_t, bits, syncable>::cgbn_t &) const [with context_t=context_t, bits=1024U, syncable=(cgbn_syncable_t)0]" 
/CGBN-master/include/cgbn/cgbn.h(132): here
            instantiation of "void cgbn_div(env_t, env_t::cgbn_t &, const env_t::cgbn_t &, const env_t::cgbn_t &) [with env_t=env_t]" 
modnumspiral.cu(140): here

I will also attach an excerpt of the .cu file (saved as TXT here) where the error occurs at line 140, more precisely when executing a **cgbn_div(bn_env, nrr, rez2, nmm);** instruction.

Note: the four basic CGBN headers, that are
#include "cgbn/cgbn.h"
#include "utility/cpu_support.h"
#include "utility/cpu_simple_bn_math.h"
#include "utility/gpu_support.h"

are already included in another header file (CUDAFUNC.h) that my main .cu file calls. They are run from there, so all inclusions are assured. And for compiling the code I use the command

**nvcc -I"/CudaSamples/cuda-samples-master/Common" -I"/root/NVIDIA_GPU_Computing_SDK/C/common/inc/" -I"/CGBN-master/include" -I"/CGBN-master/samples" -arch=sm_89 modnumspiral.cu -o modnumspiral -lcurand -lgmp -Xcompiler=-fopenmp -w
**

, thus ensuring that all inclusion requirements are met. Including the required headers and libraries works fine.
So my CGBN environment seems to be properly instantiated, I have tested CGBN and got a lot of results for BITS=256, then what is happening to the division routines if we want to use BITS=1024?

By the way, are you planning to do a new CGBN release with improved functionalities? It would be great for us to be able to work with big integer numbers on CUDA in a flawless way, as CGBN still has some important limitations. CUDA is able to provide us great speed on 4090 video cards, but the software also needs to run well and flawless.

In conclusion, I would need some help with the division errors above, when using BITS > 512 on CGBN. I am a person who loves working with great integer numbers and CGBN may be a good library for achieving this if it is tuned well.

Thanks in advance!
Peter Minea

[modnumspiral.TXT](https://github.com/NVlabs/CGBN/files/10341583/modnumspiral.TXT)

**Comment by stevvvvvvv** [comment]
Found same issue as yours. This happens when LIMBS are larger than TPI because in [dispatch_dlimbs.cu](https://github.com/NVlabs/CGBN/blob/e8b9d265c7b84077d02340b0986f3c91b2eb02fb/include/cgbn/core/dispatch_dlimbs.cu#L27) it defined four dlimbs_algs_t but only instantiated three.
